ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -973,24 +974,33 @@ public class Manager: NSObject, HTTPDownloadDelegate, UIGestureRecognizerDelegat
   
   func downloadLexicalModelPackage(string lexicalModelPackageURLString: String) -> Void {
     if let lexicalModelKMPURL = URL.init(string: lexicalModelPackageURLString) {
-      //data from URL
-      guard let data = try? Data(contentsOf: lexicalModelKMPURL) else {
-        let error = NSError(domain: "Keyman", code: 0,
-                            userInfo: [NSLocalizedDescriptionKey: "Failed to fetch lexical model KMP file"])
-        downloadFailed(forKeyboards: [], error: error)
-        return
-      }
-      //put the data at a local  file  URL
+      //determine where to put the data (local  file  URL)
       var destinationUrl = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
       destinationUrl.appendPathComponent("\(lexicalModelKMPURL.lastPathComponent).zip")
-      do {
-        //write the URL contents to a local file URL
-        try data.write(to: destinationUrl)
-        installLexicalModelPackage(downloadedPackageFile: destinationUrl)
-        log.info("lexical model download complete")
-      } catch {
-        log.error("writing or opening downloaded lexical model KMP failed")
+      //callback to handle the data downloaded
+      func lexicalModelDownloaded(data: Data?,
+                                  response: URLResponse?,
+                                  dest: URL,
+                                  error: Error?) {
+        if let error = error {
+          log.error("Failed to fetch lexical model KMP file")
+          downloadFailed(forKeyboards: [], error: error)
+        } else {
+          do {
+            try data!.write(to: dest)
+          } catch {
+            log.error("Error writing the lexical model download data: \(error)")
+          }
+          installLexicalModelPackage(downloadedPackageFile: dest)
+        }
       }
+
+      log.info("downloading lexical model from Keyman cloud: \(lexicalModelKMPURL).")
+      let task = URLSession.shared.dataTask(with: lexicalModelKMPURL) { (data, response, error) in
+        lexicalModelDownloaded(data: data, response: response, dest: destinationUrl, error: error)
+      }
```

Comment:
So, this brings things more in line with existing API calls for lists of data - great when we want to produce those lexical model list views.

What the iOS app has been using for specific file downloads, though, can best be seen [here, starting at line 535 in Manager.swift](https://github.com/keymanapp/keyman/blob/master/ios/engine/KMEI/KeymanEngine/Classes/Manager.swift#L535).
* Note the line about `request.destinationFile = Storage.active.keyboardURL(for: keyboard).path`.  This pre-assigns a destination path for the downloaded file, which our `HTTPDownloader` class will handle on your behalf.

After a block like the following:
```
downloadQueue!.addRequest(request)
downloadQueue!.run()
```

The download queue will then run asynchronously to request and deliver the specifically-requested files from online.

* Upon completing download of an individual file, the download queue then calls [line 730 of Manager.swift](https://github.com/keymanapp/keyman/blob/master/ios/engine/KMEI/KeymanEngine/Classes/Manager.swift#L737), with the relevant section starting at line 737.
  * Right now, the download queue only handles download of keyboards, so extra checks and notification types would admittedly be needed here.
* The `keyboardDownloadCompleted` notification then triggers [line 187 of KeyboardPickerViewController](https://github.com/keymanapp/keyman/blob/master/ios/engine/KMEI/KeymanEngine/Classes/LanguagePicker/KeyboardPickerViewController.swift#L187), which handles actually adding and installing the keyboard within the app.
  * In particular, note [line 232's block](https://github.com/keymanapp/keyman/blob/master/ios/engine/KMEI/KeymanEngine/Classes/LanguagePicker/KeyboardPickerViewController.swift#L232).
* That `addKeyboard` call then triggers [line 231 of Manager.swift](https://github.com/keymanapp/keyman/blob/master/ios/engine/KMEI/KeymanEngine/Classes/Manager.swift#L231), which permanently registers the downloaded keyboard for use within the app.

If you have any questions about the `HTTPDownloader` class section, which I've largely glossed over here, let me know - I'm the one who wrote it.  You should be safe ignoring its internals, though there's a chance some tweaks may be needed to the `HTTPDownloadRequest` if we want to make a new download "type."

----

That said... this chain can get a little involved, and would be a fair bit of work to implement.  @mcdurdin, would it be reasonable to accept things as they are for now and make an issue for conversion to this specified pattern when the "lexical model picker" GUIs are implemented?

Reply:
I completely agree. Both with the further improvement (I really didn't like that I get  a whole  Data object in memory and then write it to the destination) and with putting off perfection until later.